packages/react-scheduler/src/ReactScheduler.js

```diff
@@ -200,35 +200,78 @@ if (!ExecutionEnvironment.canUseDOM) {
     callback: (deadline: Deadline) => void,
     options?: {timeout: number},
   ): number {
-    if (scheduledRICCallback !== null) {
-      // Handling multiple callbacks:
-      // For now we implement the behavior expected when the callbacks are
-      // serial updates, such that each update relies on the previous ones
-      // having been called before it runs.
-      frameDeadlineObject.didTimeout =
-        (timeoutTime !== -1) && (timeoutTime <= now());
-      scheduledRICCallback(frameDeadlineObject);
+    // Handling multiple callbacks:
+    // For now we implement the behavior expected when the callbacks are
+    // serial updates, such that each update relies on the previous ones
+    // having been called before it runs.
+    // So we call anything in the queue before the latest callback
+
+    let previousCallback;
+    let timeoutTimeFromPreviousCallback;
+    if (scheduledCallback !== null) {
+      // If we have previous callback, save it and handle it below
+      timeoutTimeFromPreviousCallback = timeoutTime;
+      previousCallback = scheduledCallback;
     }
-    scheduledRICCallback = callback;
+    // Then set up the next callback, and update timeoutTime
+    scheduledCallback = callback;
     if (options != null && typeof options.timeout === 'number') {
       timeoutTime = now() + options.timeout;
     } else {
       timeoutTime = -1;
     }
+    // If we have previousCallback, call it. This may trigger recursion.
+    if (
+      previousCallback &&
```

Comment:
These checks are wasteful, since we already checked if `scheduleCallback` is null above. This block should be moved into the earlier one:

- Check if there's an already scheduled callback
  - If so, flush it
- Schedule the incoming callback